lib/grandpa/vote_message_test.go

```diff
@@ -0,0 +1,186 @@
+	err := equivocationVote.Set(types.PreVoteEquivocation(grandpaEquivocation))
+	require.NoError(t, err)
+
+	equivocationProof := types.GrandpaEquivocationProof{
+		SetID:        uint64(1),
+		Equivocation: *equivocationVote,
+	}
+
+	ctrl := gomock.NewController(t)
+	mockGrandpaStateSetIDErr := NewMockGrandpaState(ctrl)
+	mockGrandpaStateSetIDErr.EXPECT().GetCurrentSetID().Return(uint64(0), errTestError)
+
+	mockGrandpaStateRoundErr := NewMockGrandpaState(ctrl)
+	mockGrandpaStateRoundErr.EXPECT().GetCurrentSetID().Return(uint64(1), nil)
+	mockGrandpaStateRoundErr.EXPECT().GetLatestRound().Return(uint64(0), errTestError)
+
+	mockGrandpaStateOk := NewMockGrandpaState(ctrl)
+	mockGrandpaStateOk.EXPECT().GetCurrentSetID().Return(uint64(1), nil).Times(4)
+	mockGrandpaStateOk.EXPECT().GetLatestRound().Return(uint64(1), nil).Times(4)
+
+	mockRuntimeInstanceGenerateProofErr := NewMockRuntimeInstance(ctrl)
+	mockRuntimeInstanceGenerateProofErr.EXPECT().GrandpaGenerateKeyOwnershipProof(uint64(1), testAuthorityID).
+		Return(types.OpaqueKeyOwnershipProof{}, errTestError)
+
+	mockRuntimeInstanceReportEquivocationErr := NewMockRuntimeInstance(ctrl)
+	mockRuntimeInstanceReportEquivocationErr.EXPECT().GrandpaGenerateKeyOwnershipProof(uint64(1), testAuthorityID).
+		Return(keyOwnershipProof, nil)
+	mockRuntimeInstanceReportEquivocationErr.EXPECT().
+		GrandpaSubmitReportEquivocationUnsignedExtrinsic(equivocationProof, keyOwnershipProof).
+		Return(errTestError)
+
+	mockRuntimeInstanceOk := NewMockRuntimeInstance(ctrl)
+	mockRuntimeInstanceOk.EXPECT().GrandpaGenerateKeyOwnershipProof(uint64(1), testAuthorityID).
+		Return(keyOwnershipProof, nil)
+	mockRuntimeInstanceOk.EXPECT().
+		GrandpaSubmitReportEquivocationUnsignedExtrinsic(equivocationProof, keyOwnershipProof).
+		Return(nil)
+
+	mockBlockStateGetRuntimeErr := NewMockBlockState(ctrl)
+	mockBlockStateGetRuntimeErr.EXPECT().BestBlockHash().Return(testDummyHash)
+	mockBlockStateGetRuntimeErr.EXPECT().GetRuntime(testDummyHash).Return(nil, errTestError)
+
+	mockBlockStateGenerateProofErr := NewMockBlockState(ctrl)
+	mockBlockStateGenerateProofErr.EXPECT().BestBlockHash().Return(testDummyHash)
+	mockBlockStateGenerateProofErr.EXPECT().GetRuntime(testDummyHash).
+		Return(mockRuntimeInstanceGenerateProofErr, nil)
+
+	mockBlockStateReportEquivocationErr := NewMockBlockState(ctrl)
+	mockBlockStateReportEquivocationErr.EXPECT().BestBlockHash().Return(testDummyHash)
+	mockBlockStateReportEquivocationErr.EXPECT().GetRuntime(testDummyHash).
+		Return(mockRuntimeInstanceReportEquivocationErr, nil)
+
+	mockBlockStateOk := NewMockBlockState(ctrl)
+	mockBlockStateOk.EXPECT().BestBlockHash().Return(testDummyHash)
+	mockBlockStateOk.EXPECT().GetRuntime(testDummyHash).Return(mockRuntimeInstanceOk, nil)
+
+	type args struct {
+		stage        Subround
+		existingVote *SignedVote
+		currentVote  *SignedVote
+	}
+	tests := []struct {
+		name      string
+		service   *Service
+		args      args
+		expErr    error
+		expErrMsg string
```

Comment:
define these mocks for each test case using a functional field in the test struct:

```go
serviceBuilder func(ctrl *gomock.Controller) *Service
```

And configure the mocks for each test case.

The test loop body can then do

```go
ctrl := gomock.NewController(t)
service := tt.serviceBuilder(ctrl)
```

That way a reader doesn't have to jump up and down, everything is in the test case definition block (and also controller isn't shared across parallel subtests)

Reply:
Really like this!

Reply:
Credits to Tim a few months back 🎖️